jumanji/environments/routing/sokoban/env.py

```diff
@@ -0,0 +1,587 @@
+            time_limit: int, max steps for the environment ,defaults to 120.
+            viewer: 'Viewer' object, used to render the environment.
+            If not provided, defaults to`BoxViewer`.
+        """
+
+        self.num_rows = GRID_SIZE
+        self.num_cols = GRID_SIZE
+        self.shape = (self.num_rows, self.num_cols)
+        self.time_limit = time_limit
+        self.generator = generator or HuggingFaceDeepMindGenerator(
+            "unfiltered-train",
+            proportion_of_files=1,
+        )
+        self._viewer = viewer or BoxViewer(
+            name="Sokoban",
+            grid_combine=self.grid_combine,
+        )
+        self.reward_fn = reward_fn or DenseReward()
+
+    def __repr__(self) -> str:
+        """
+        Returns a printable representation of the Sokoban environment.
+
+        Returns:
+            str: A string representation of the Sokoban environment.
+        """
+        return "\n".join(
+            [
+                "Bokoban environment:",
+                f" - num_rows: {self.num_rows}",
+                f" - num_cols: {self.num_cols}",
+                f" - time_limit: {self.time_limit}",
+                f" - generator: {self.generator}",
+            ]
+        )
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        """
+        Resets the environment by calling the instance generator for a
+        new instance.
+
+        Args:
+            key: random key used to sample new Sokoban problem.
+
+        Returns:
+            state: `State` object corresponding to the new state of the
+            environment after a reset.
+            timestep: `TimeStep` object corresponding the first timestep
+            returned by the environment after a reset.
+        """
+
+        generator_key, key = jax.random.split(key)
+
+        state = self.generator(generator_key)
+
+        timestep = restart(
+            self._state_to_observation(state),
+            extras=self._get_extras(state),
+        )
+
+        return state, timestep
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        """
+        Executes one timestep of the environment's dynamics.
+
+        Args:
+            state: 'State' object representing the current state of the
+            environment.
+            action: Array (int32) of shape ().
+                - 0: move up.
+                - 1: move down.
+                - 2: move left.
+                - 3: move right.
+
+        Returns:
+            state, timestep: next state of the environment and timestep to be
+            observed.
+        """
+
+        # switch to noop if action will have no impact on variable grid
+        action = self.detect_noop_action(
+            state.variable_grid, state.fixed_grid, action, state.agent_location
+        )
+
+        next_variable_grid, next_agent_location = jax.lax.cond(
+            jnp.all(action == -1),
```

Comment:
Maybe we should say somewhere that -1 is a no-op?

Reply:
Maybe add this to constants.py?